core/desugarMatching.ml

```diff
@@ -54,6 +62,7 @@ object ((self : 'self_type))
           (* bind the arguments with unique var name *)
           let name_list = List.map (fun pats -> List.map (fun pat -> (pat, Utility.gensym())) pats) patterns in
           let switch_tuple = List.map (fun (_, name) -> with_pos (Var name)) (List.flatten name_list) in
+          nullary_guard switch_tuple pos;
```

Comment:
This is somewhat nitpick, but we might as well move this check up such that it is executed right after the flag check (`pattern_matching_sugar_guard`). It'd require the implementation of `nullary_guard` to match on `[[]]` rather than `[]`.

```suggestion
          nullary_guard patterns pos;
```